### DISCUSSION TITLE
issue: inputs are ignored if reset() is called in effect after first render

### DISCUSSION THREAD

**Discussion by d-s-x** [react-hook-form/react-hook-form#9478] [Q&A]
### Version Number

7.40.0

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-basic-34hiro?file=/src/index.js
https://codesandbox.io/s/react-hook-form-basic-2-d0wxdg?file=/src/index.js

### Steps to reproduce

See below

### Expected behaviour

I have an application with "smart" `Form` and `Field` components. But I have to call `useForm()` outside of the `Form` component so I have access to `watch`, `control`, etc.

In the sample application I have a list of users, and a form to edit them. Select user, change something, submit, nothing fancy.
The `Form` component calls `reset()` every time the selection changes. (This behavior mimics Formik, my app used it previously.) Somehow the first reset breaks hook-form internal state, but second one "fixes" it.

1. Go to https://codesandbox.io/s/react-hook-form-basic-34hiro?file=/src/index.js
  Observe:
* there is a "list" of users ontop
* first one is selected for editing
* it's details are pre-filled in the form below ("John"/"Doe"/"")
2. Click into "First Name" field and type "atan"
3. Click Submit
Expected: alert should display `firstName: Johnatan`.
Actual: `firstName: John` - no changes are propagated, inputs are ignored.

4. Click "Jane" button to select second user (which calls reset() internally), repeat steps 2,3
Observe: input is no longer ignored

NOTE: My original though was that React development mode might be the cause but the issue reproduces on production build too.

Also, in this example I had to pass initial values both into `useForm({defaultValues: ...})` and `<Form initialValues={...}`
which looks superfluous. I think it should work even if I pass initialValues only to my `Form` component as it will call `reset()` anyway which I expect should override any default values passed to `useForm()`.
But in this case the inputs are empty.

1. Go to https://codesandbox.io/s/react-hook-form-basic-2-d0wxdg?file=/src/index.js
(this is exactly the same code as above, except `useForm({defaultValues: initialValues})` is replaced with just `useForm()`)
Expected: form inputs are filled with first user details ("John"/"Doe"/"") as this user is selected automatically
Actual: the form inputs are empty
2. Type "Foo" into "First Name" field
3. Click "Submit"
Expected: alert should display `firstName: Foo, lastName: ""`.
Actual: `firstName: John, lastName: Doe`

4. Click "Jane" button to select second user (which calls reset() internally), repeat steps 2,3
Observe: input is no longer ignored -- like in previous use case

### What browsers are you seeing the problem on?

Firefox, Chrome

### Relevant log output

there are no errors nor warnings in the log

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
https://react-hook-form.com/api/useform/reset

> Avoid calling reset before useForm's useEffect is invoked, this is because useForm's subscription needs to be ready before reset can send a signal to flush form state update.

![Screenshot 2022-11-30 at 9 12 40 pm](https://user-images.githubusercontent.com/10513364/204768948-dbaf1414-96e9-4e2e-980c-c0aa6ea6230e.png)

Move the `reset` at `useForm`

**Comment by d-s-x** [comment]
I didn't mention that in OP, but I actually have lots of forms and I'm not working on the code alone.

When a person is looking at `useForm({defaultProps})` it is not obvious that changes to `defaultProps` are ignored and you need to call `reset()`. I can't possibly convey that to all current and future collaborators. And they definitely won't read all the documentation for all the frameworks we are using.

Not to mention that adding reset in many places (compared to one call in shared component) is code duplication.

> Avoid calling reset before useForm's useEffect is invoked, this is because useForm's subscription needs to be ready before reset can send a signal to flush form state update.

I understand you might not want to fix that limitation but maybe you can detect that use case and print a warning in development mode? (like React does for lots of things)

But from API point of view it makes no sense. If I created a form and called reset() the state must be reset. Having to track number of react re-renders since first call to useForm() before using reset() is so error-prone.

I think the issue can be solved relatively easy.
For example, if you know that `reset()` will not work properly before certain useEffect() was invoked you can introduce two implementations for reset.
One that works just like before and another that writes to temporary state.
When effect is invoked it will check temporary state and apply necessary adjustments, and switch reset() implementation from second to first.

That maybe easier to illustrate as pseudocode
```
let initialized = false;
let resetTmp = null;
function reset1(val) {
  // original reset code
}
function reset2(val) {
  resetTmp = { params: val }
}
function reset(val) {
  (initialized ? reset1 : reset2)(val)
}
useEffect(() => {
  // original effect code
  if (resetTmp) {
    applyDelayedReset(resetTmp.params)
  }
  initialized = true;
}
```

**Reply by d-s-x** [reply]
@bluebill1049 so, what's your take on that? Do you agree a fix or a warning is welcome. Or do you think that reset is working as intended?

**Comment by leapful** [comment]
@d-s-x 

This is a common issue when working with uncontrolled React component, not just react-hook-form library because method `register` will take advantages on `ref` to handle form field behaviors and avoid re-rendering, therefore, it depends on DOM element to be ready to be manipulated. You cannot reset a field that is not available yet.

The workaround is simple too, just use `setTimeout` to put `reset` on top of event queue.

Working demo: https://codesandbox.io/s/react-hook-form-basic-forked-gynp8z?file=/src/index.js

**Reply by d-s-x** [reply]
@leapful, thanks for suggestion but timeout-based hacks are the worst.

First of all, this workaround still requires to provide initial values to both useForm() call, and Form component even though Form's contact assumes it will handle the value on its own. But if I remove the defaultValue parameter from useForm then the form is rendered empty and "blinks" into filled state. (BTW, it means we have unnecessary render here.) Which means I can not omit defaultValue from useForm. Which means fist reset() call is not doing anything useful and the same result can be achieved by tweaking Form component's logic to just skip first reset entirely.

Second, setTimeout hacks have a snowball effect. I worked with old ag-grid approx. 5 years ago and some API calls were not returning what they were supposed to return unsell I skip a frame (I don;t remeber what it was exactly but something like adding a column,reading back list of columns, the new column wasn't there). I ended up with double setTimeout(()=>setTimeout(...)) but they fixed their mistakes in newer versions. Not to mention how "fun" and "intuitive" it was to work with.

In other words, my point is that most obvious way to solve a task should be the correct one. And most obvious way to reset the state when initial values changes is **to add a flag to useForm so it treats defaultValues reactively**. Especially considering that other form management solutions also have that feature (like in Formik, but I don't like to use it for other reasons).

**Comment by leapful** [comment]
@d-s-x 

It depends on your code structure to apply the best workaround. The problem is from the order of rendering between parent and child component, where the default values must be set before field elements are available and populate data after it has been initialized.

Using `React.useLayoutEffect` and `React.useEffect` on the parent component will do the trick.

Working demo without setTimeout: https://codesandbox.io/s/react-hook-form-basic-forked-v6e2ko?file=/src/index.js
